compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -344,17 +344,32 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
     ) -> ProcessResult<PendingPredicateObligation<'tcx>, FulfillmentErrorCode<'tcx>> {
         pending_obligation.stalled_on.truncate(0);
 
+        let infcx = self.selcx.infcx();
         let obligation = &mut pending_obligation.obligation;
 
-        if obligation.predicate.has_infer_types_or_consts() {
+        debug!(?obligation, "process_obligation pre-resolve");
+
+        if obligation.predicate.has_projections() {
+            let mut obligations = Vec::new();
+            let predicate = crate::traits::project::try_normalize_with_depth_to(
+                self.selcx,
+                obligation.param_env,
+                obligation.cause.clone(),
+                obligation.recursion_depth + 1,
+                obligation.predicate,
+                &mut obligations,
+            );
+            if predicate != obligation.predicate {
+                obligations.push(obligation.with(predicate));
+                return ProcessResult::Changed(mk_pending(obligations));
+            }
+        } else if obligation.predicate.has_infer_types_or_consts() {
```

Comment:
Just fyi this now skips the inference var resolution if there is a projection but we can't normalize.

Reply:
I was thinking about that just now... Do you want me to revert and we can re-queue the test, or should we just do perf on whatever's currently queued?